zb-db/src/main/java/io/zeebe/db/impl/rocksdb/ZeebeRocksDbFactory.java

```diff
@@ -55,84 +68,189 @@ private ZeebeRocksDbFactory(
+                    + "See RocksDB's cf_options.h and options_helper.cc for available keys and values.",
+                columnFamilyOptionProps, userProvidedColumnFamilyOptions));
+      }
+      return columnFamilyOptions;
     }
-    return columnFamilyOptions;
+
+    return createDefaultColumnFamilyOptions(closeables);
+  }
+
+  private ColumnFamilyOptions createDefaultColumnFamilyOptions(
+      final List<AutoCloseable> closeables) {
+    final var columnFamilyOptions = new ColumnFamilyOptions();
+
+    // given
+    final long totalMemoryBudget = 512 * 1024 * 1024L; // make this configurable
+    // recommended by RocksDB, but we could tweak it; keep in mind we're also caching the indexes
+    // and filters into the block cache, so we don't need to account for more memory there
+    final long blockCacheMemory = totalMemoryBudget / 3;
+    // flushing the memtables is done asynchronously, so there may be multiple memtables in memory,
+    // although only a single one is writable. once we have too many memtables, writes will stop.
+    // since prefix iteration is our bread n butter, we will build an additional filter for each
+    // memtable which takes a bit of memory which must be accounted for from the memtable's memory
+    final int maxConcurrentMemtableCount = 6;
+    final double memtablePrefixFilterMemory = 0.15;
+    final long memtableMemory =
+        Math.round(
+            ((totalMemoryBudget - blockCacheMemory) / (double) maxConcurrentMemtableCount)
+                * (1 - memtablePrefixFilterMemory));
+
+    final TableFormatConfig tableConfig = createTableFormatConfig(closeables, blockCacheMemory);
+
+    return columnFamilyOptions
+        // prefix seek must be fast, so allocate an extra 10% of a single memtable budget to create
+        // a filter for each memtable, allowing us to skip them if possible
+        .useFixedLengthPrefixExtractor(Long.BYTES)
+        .setMemtablePrefixBloomSizeRatio(memtablePrefixFilterMemory)
+        // memtables
+        // merge at least 3 memtables per L0 file, otherwise all memtables are flushed as individual
+        // files
+        .setMinWriteBufferNumberToMerge(Math.min(3, maxConcurrentMemtableCount))
+        .setMaxWriteBufferNumberToMaintain(maxConcurrentMemtableCount)
+        .setMaxWriteBufferNumber(maxConcurrentMemtableCount)
+        .setWriteBufferSize(memtableMemory)
+        // compaction
+        .setLevelCompactionDynamicLevelBytes(true)
+        .setCompactionPriority(CompactionPriority.OldestSmallestSeqFirst)
+        .setCompactionStyle(CompactionStyle.LEVEL)
+        // L-0 means immediately flushed memtables
+        .setLevel0FileNumCompactionTrigger(maxConcurrentMemtableCount)
+        .setLevel0SlowdownWritesTrigger(
+            maxConcurrentMemtableCount + (maxConcurrentMemtableCount / 2))
+        .setLevel0StopWritesTrigger(maxConcurrentMemtableCount * 2)
+        // configure 4 levels: L1 = 32mb, L2 = 320mb, L3 = 3.2Gb, L4 >= 3.2Gb
```

Comment:
If we want to support bigger states, we'll probably want to add levels. Can we add a follow up issue for this?

Reply:
Feel free to create one :D

Reply:
https://github.com/zeebe-io/zeebe/issues/6179

Reply:
Thanks @npepinpe :+1: